### PR TITLE
Updating order with a custom controller action

### PR DESCRIPTION
**Pull request by Zales0123**
This PR does not provide any additional functionalities but refactors the current logic to get rid of the decorated resource update handler and artificial additional state machine transition 🖖

### CODE REVIEW

File: psalm.xml [hunk 1]
```diff
@@ -15,6 +15,7 @@
         <directory name="tests"/>
         <ignoreFiles>
             <directory name="tests/Application"/>
+            <directory name="tests/Functional"/>
```
Comment (loevgaard) [inline psalm.xml:18]: I try to avoid this, but I guess you have your reasons? :D

File: src/EventSubscriber/AddEditOrderActionSubscriber.php [hunk 2]
```diff
@@ -9,19 +9,14 @@
 use Sylius\Component\Grid\Event\GridDefinitionConverterEvent;
 use Symfony\Component\EventDispatcher\EventSubscriberInterface;
 
-/**
```
Comment (loevgaard) [inline src/EventSubscriber/AddEditOrderActionSubscriber.php:12]: What's wrong with this? Too obvious? :)

File: src/Provider/UpdatedOrderProvider.php [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+use Sylius\Bundle\OrderBundle\Form\Type\OrderType;
+use Sylius\Component\Core\Model\OrderInterface;
+use Symfony\Component\Form\FormFactoryInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Webmozart\Assert\Assert;
+
+final class UpdatedOrderProvider
+{
+    public function __construct(private FormFactoryInterface $formFactory)
```
Comment (loevgaard) [inline src/Provider/UpdatedOrderProvider.php]:
```suggestion
    public function __construct(private readonly FormFactoryInterface $formFactory)
```

Just a small 'best practice' I always do

File: src/Provider/OldOrderProvider.php [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    ) {
+    }
+
+    public function provide(int $orderId): OrderInterface
+    {
+        /** @var OrderInterface|null $order */
+        $order = $this->orderRepository->find($orderId);
+        Assert::isInstanceOf($order, OrderInterface::class);
+
+        $this->orderInventoryOperator->cancel($order);
```
Comment (loevgaard) [inline src/Provider/OldOrderProvider.php]: Taking the name of the service into consideration I wouldn't expect this call to happen here

File: src/Provider/OldOrderProvider.php [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+declare(strict_types=1);
+
+namespace Setono\SyliusOrderEditPlugin\Provider;
+
+use Sylius\Component\Core\Inventory\Operator\OrderInventoryOperatorInterface;
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Repository\OrderRepositoryInterface;
+use Webmozart\Assert\Assert;
+
+final class OldOrderProvider
```
Comment (loevgaard) [inline src/Provider/OldOrderProvider.php]: Missing interface.

What are your thoughts behind having a whole service for fetching an order from the repository?

File: src/Resources/views/admin/order/update/_order_items.html.twig [hunk 6]
```diff
@@ -28,7 +28,7 @@
                 {% set orderItem = itemForm.vars.data %}
                 <tr>
                     <td>{{ form_row(itemForm.quantity) }}</td>
-                    <td>{{ orderItem.productName }}</td> {# todo add variant #}
+                    <td>{{ form_widget(itemForm.variant) }}</td>
```
Comment (loevgaard) [inline src/Resources/views/admin/order/update/_order_items.html.twig]: Why did you change this to `form_widget`. Is it so that _existing_ order lines can be edited? For 'safety reasons' I wouldn't do that myself. What do you think?

File: tests/Functional/OrderUpdateTest.php [hunk 7]
```diff
@@ -0,0 +1,158 @@
+<?php
```
Comment (loevgaard) [inline tests/Functional/OrderUpdateTest.php:1]: ❤️ ❤️ ❤️ ❤️ ❤️

File: composer.json [hunk 8]
```diff
@@ -18,11 +19,13 @@
         "sylius/resource-bundle": "^1.6",
         "symfony/config": "^5.4 || ^6.4 || ^7.0",
         "symfony/dependency-injection": "^5.4 || ^6.4 || ^7.0",
+        "symfony/doctrine-bridge": "^5.4 || ^6.4 || ^7.0",
```
Comment (loevgaard) [inline composer.json]:
```suggestion
        "symfony/doctrine-bridge": "^5.4 || ^6.3 || ^7.0",
```

`symfony/doctrine-bridge` v6.4 removes `doctrine/persistence` v2 from its deps, and this has a lot of implications for Sylius related packages

File: composer.json [hunk 9]
```diff
@@ -34,10 +37,12 @@
         "jms/serializer-bundle": "^4.2",
         "lexik/jwt-authentication-bundle": "^2.17",
         "matthiasnoback/symfony-dependency-injection-test": "^4.3 || ^5.1",
+        "phpspec/prophecy-phpunit": "^2.2",
         "phpunit/phpunit": "^9.6.17",
         "psalm/plugin-phpunit": "^0.18.4",
         "setono/code-quality-pack": "^2.7",
         "sylius/sylius": "~1.12.13",
+        "symfony/browser-kit": "^5.4",
```
Comment (loevgaard) [inline composer.json]:
```suggestion
        "symfony/browser-kit": "^5.4 || ^6.4 || ^7.0",
```

File: composer.json [hunk 10]
```diff
@@ -80,6 +85,7 @@
         "analyse": "psalm",
         "check-style": "ecs check",
         "fix-style": "ecs check --fix",
-        "phpunit": "phpunit"
+        "functional-tests": "vendor/bin/phpunit tests/Functional/",
+        "unit-tests": "vendor/bin/phpunit tests/Unit/"
```
Comment (loevgaard) [inline composer.json]:
```suggestion
        "functional-tests": "phpunit tests/Functional/",
        "unit-tests": "phpunit tests/Unit/"
```

File: src/Controller/EditOrderAction.php [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    private function addFlashAndRedirect(
+        string $type,
+        string $message,
+        string $route,
+        int $orderId,
+    ): RedirectResponse {
+        $session = $this->requestStack->getSession();
+        /** @var FlashBagInterface $flashBag */
```
Comment (loevgaard) [inline src/Controller/EditOrderAction.php]: It's annoying, but I try to avoid phpdocs like these. Mostly to make things as robust as possible. Maybe it's a false notion...

For this example (and when I support SF5.4) I just do

```php 
if($session instance of \Symfony\Component\HttpFoundation\Session\Session) {
    $session->getFlashbag()->add(...);
}
```

File: src/Exception/NewOrderWrongTotalException.php [hunk 12]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Setono\SyliusOrderEditPlugin\Exception;
+
+final class NewOrderWrongTotalException extends \RuntimeException
+{
+    public static function occur(): self
```
Comment (loevgaard) [inline src/Exception/NewOrderWrongTotalException.php]: Is this some kind of pattern?

Not sure what it improves over `new NewOrderWrongTotalException`.

If I do something like this, I would do something like this instead:

```php
public static function assert(int $oldTotal, int $newTotal): void
{
    if($oldTotal < $newTotal) {
        throw new self('New order total is greater than the initial order total');
    }
}
```

File: src/Processor/UpdatedOrderProcessor.php [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Order\Processor\OrderProcessorInterface;
+
+final class UpdatedOrderProcessor
+{
+    public function __construct(
+        private OrderProcessorInterface $orderProcessor,
+        private OrderInventoryOperatorInterface $orderInventoryOperator,
+        private OrderProcessorInterface $afterCheckoutOrderPaymentProcessor,
+        private EntityManagerInterface $entityManager,
```
Comment (loevgaard) [inline src/Processor/UpdatedOrderProcessor.php]: I never inject the entity manager. I use this library for these purposes: https://github.com/Setono/doctrine-orm-trait

File: src/Provider/OldOrderProvider.php (hunk 4 above)
Comment (loevgaard) [reply src/Provider/OldOrderProvider.php]: Also; the reason I did it manually in the old code was that I am not sure that it's safe to call `cancel`. The `cancel` method depends on the payment state. So in a scenario where the order is paid, you will put items back on hand and afterwards put them on hold, which I suppose can mess up the inventory tracking

File: src/Controller/EditOrderAction.php [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+        try {
+            $this->updatedOrderProcessor->process($initialTotal, $resource);
+        } catch (NewOrderWrongTotalException) {
+            return $this->addFlashAndRedirect(
+                'error',
+                'setono_sylius_order_edit.error.order_update',
+                'sylius_admin_order_update',
+                $orderId,
+            );
+        }
```
Comment (loevgaard) [inline src/Controller/EditOrderAction.php:58]: We need to handle other exceptions also, don't we?

File: src/Processor/UpdatedOrderProcessor.php [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+
+namespace Setono\SyliusOrderEditPlugin\Processor;
+
+use Doctrine\ORM\EntityManagerInterface;
+use Setono\SyliusOrderEditPlugin\Exception\NewOrderWrongTotalException;
+use Sylius\Component\Core\Inventory\Operator\OrderInventoryOperatorInterface;
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Order\Processor\OrderProcessorInterface;
+
+final class UpdatedOrderProcessor
```
Comment (loevgaard) [inline src/Processor/UpdatedOrderProcessor.php]: Missing interface

File: src/Provider/UpdatedOrderProvider.php [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+
+namespace Setono\SyliusOrderEditPlugin\Provider;
+
+use Sylius\Bundle\OrderBundle\Form\Type\OrderType;
+use Sylius\Component\Core\Model\OrderInterface;
+use Symfony\Component\Form\FormFactoryInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Webmozart\Assert\Assert;
+
+final class UpdatedOrderProvider
```
Comment (loevgaard) [inline src/Provider/UpdatedOrderProvider.php]: Missing interface

File: src/Resources/translations/flashes.en.yaml [hunk 17]
```diff
@@ -0,0 +1,5 @@
+setono_sylius_order_edit:
+    error:
+        order_update: 'Order should not have bigger total than before editing'
+    success:
+        order_update: 'Order has been successfully updated'
```
Comment (loevgaard) [inline src/Resources/translations/flashes.en.yaml]: Is this a normal structure? I would have expected something like this:

```suggestion
setono_sylius_order_edit:
    order_update:
        error: 'Order should not have bigger total than before editing'
        success: 'Order has been successfully updated'
```

File: tests/Application/config/routes/sylius_admin.yaml [hunk 18]
```diff
@@ -1,3 +1,6 @@
 sylius_admin:
     resource: "@SyliusAdminBundle/Resources/config/routing.yml"
     prefix: '/%sylius_admin.path_name%'
+
+setono_edit_order:
```
Comment (loevgaard) [inline tests/Application/config/routes/sylius_admin.yaml]: Should be in its own file: `setono_sylius_order_edit.yaml`. Just to keep conventions

File: src/Provider/UpdatedOrderProvider.php [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+{
+    public function __construct(private FormFactoryInterface $formFactory)
+    {
+    }
+
+    public function fromRequest(OrderInterface $order, Request $request): OrderInterface
+    {
+        $form = $this->formFactory->create(OrderType::class, $order, ['validation_groups' => 'sylius']);
+        $form->handleRequest($request);
```
Comment (loevgaard) [inline src/Provider/UpdatedOrderProvider.php:29]: $form->isSubmitted() and $form->isValid() should be called somewhere I guess

File: src/Resources/config/services/order_processing.xml [hunk 20]
```diff
@@ -8,5 +8,21 @@
             <argument type="service" id="setono_sylius_order_edit.order_processing.order_payment_processor.inner"/>
             <argument type="service" id="request_stack"/>
         </service>
+
+        <service id="Setono\SyliusOrderEditPlugin\Provider\OldOrderProvider">
```
Comment (loevgaard) [inline src/Resources/config/services/order_processing.xml]: If we change service ids to fqcns it should be consistent. However, I don't know why I haven't switched to that yet. Probably because there are still so many services with those snake cased service names

File: psalm.xml (hunk 1 above)
Comment (Zales0123) [reply psalm.xml:18]: Psalm is complaining about tons of stuff like nullable getters or too generatic type hints - things that could be important in the application code but that are just  irrelevant in the tests, where we have 150% over data.
Still, probably skipping the whole directory is a little exaggeration 💃 I will try to be more precise 🖖

File: src/Resources/config/routes/admin.yaml [hunk 21]
```diff
@@ -1 +1,5 @@
-# TODO Add your admin routes here
\ No newline at end of file
+setono_sylius_order_edit_admin_update:
+    path: /orders/{id}/update-and-processs
```
Comment (loevgaard) [inline src/Resources/config/routes/admin.yaml]: Add requirements for id

File: src/Controller/EditOrderAction.php [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+    public function __construct(
+        private OldOrderProvider $oldOrderProvider,
+        private UrlGeneratorInterface $router,
+        private RequestStack $requestStack,
+        private UpdatedOrderProvider $updatedOrderProvider,
+        private UpdatedOrderProcessor $updatedOrderProcessor,
+    ) {
+    }
+
+    public function __invoke(Request $request): Response
```
Comment (loevgaard) [inline src/Controller/EditOrderAction.php]:
```suggestion
    public function __invoke(Request $request, int $id): Response
```

File: src/Controller/EditOrderAction.php (hunk 14 above)
Comment (Zales0123) [reply src/Controller/EditOrderAction.php:58]: We can probably catch an `\Exception` and throw some generic error flash, indeed 👍

File: src/Provider/UpdatedOrderProvider.php [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+use Symfony\Component\HttpFoundation\Request;
+use Webmozart\Assert\Assert;
+
+final class UpdatedOrderProvider
+{
+    public function __construct(private FormFactoryInterface $formFactory)
+    {
+    }
+
+    public function fromRequest(OrderInterface $order, Request $request): OrderInterface
```
Comment (loevgaard) [inline src/Provider/UpdatedOrderProvider.php]: I find the name of the class and this method a bit misleading. The input is the same as output (i.e. the same order object)

File: src/EventSubscriber/AddEditOrderActionSubscriber.php (hunk 2 above)
Comment (Zales0123) [reply src/EventSubscriber/AddEditOrderActionSubscriber.php:12]: Imo yest, what else can the `AddEditOrderActionSubscriber` do? 😄

File: src/Controller/EditOrderAction.php [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+    {
+        $orderId = (int) $request->attributes->get('id');
+        $order = $this->oldOrderProvider->provide($orderId);
+
+        $initialTotal = $order->getTotal();
+        $resource = $this->updatedOrderProvider->fromRequest($order, $request);
```
Comment (loevgaard) [inline src/Controller/EditOrderAction.php]: As mentioned elsewhere I find it odd that `$resource === $order`

File: src/Exception/NewOrderWrongTotalException.php (hunk 12 above)
Comment (Zales0123) [reply src/Exception/NewOrderWrongTotalException.php]: A matter of taste, I believe 🖖 Can for sure switch to `new ...` if you prefer. Regarding the if in exception, I would like to keep it in a service that is responsible for the application logic.

File: src/Processor/UpdatedOrderProcessor.php (hunk 15 above)
Comment (Zales0123) [reply src/Processor/UpdatedOrderProcessor.php]: I was not sure about your approach for interfaces - and probably I spent too much time lastly on the end-applications not on the packages 😅 I will add one 🫡

File: src/Processor/UpdatedOrderProcessor.php [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+        $this->orderProcessor->process($newOrder);
+        $this->afterCheckoutOrderPaymentProcessor->process($newOrder);
+        $newOrder->setState(OrderInterface::STATE_NEW);
+        $this->orderInventoryOperator->hold($newOrder);
+
+        if ($initialTotal < $newOrder->getTotal()) {
+            throw NewOrderWrongTotalException::occur();
+        }
+
+        $this->entityManager->flush();
```
Comment (loevgaard) [inline src/Processor/UpdatedOrderProcessor.php]: I think these lines belong in the controller or somewhere else. If they were removed, you could also remove the `$initialTotal` from the signature which I think would also make more sense

File: src/Processor/UpdatedOrderProcessor.php (hunk 13 above)
Comment (Zales0123) [reply src/Processor/UpdatedOrderProcessor.php]: Interesting - any particular reason for such an approach?

File: psalm.xml (hunk 1 above)
Comment (loevgaard) [reply psalm.xml:18]: Yeah, I figured it out when I read your tests. It's fine :D

File: src/Provider/OldOrderProvider.php (hunk 4 above)
Comment (Zales0123) [reply src/Provider/OldOrderProvider.php]: I totally agree with that comment - but, as it's mentioned in the PR description, this is just a refactor of the current logic, so we would be able to cover such a case later. And that's also the reason for having the separate service for that, so we can prepare the order to be updated in one place :dancer:

Anyway, I agree there is a naming issue here. Will try to figure out something better 🖖

File: src/Resources/config/services/order_processing.xml (hunk 20 above)
Comment (Zales0123) [reply src/Resources/config/services/order_processing.xml]: I got used to using FQCN already for a long time. Anyway, it's up to you, it's your library 😄 It's only a matter of convention, doesn't change anything 🚀

File: src/Resources/translations/flashes.en.yaml (hunk 17 above)
Comment (Zales0123) [reply src/Resources/translations/flashes.en.yaml]: I probably had a sudden attack of stupidity when writing this XD totally agreed 🖖

File: src/Resources/views/admin/order/update/_order_items.html.twig (hunk 6 above)
Comment (Zales0123) [reply src/Resources/views/admin/order/update/_order_items.html.twig]: I was 100% sure we want to do that :dancer: I understand we do not :) 
Anyway, we should probably have a disabled field with variant here, as in other case the form nulls the value on the order item.

File: src/EventSubscriber/AddEditOrderActionSubscriber.php (hunk 2 above)
Comment (loevgaard) [reply src/EventSubscriber/AddEditOrderActionSubscriber.php:12]: Haha, you're right

File: src/Resources/views/admin/order/update/_order_items.html.twig (hunk 6 above)
Comment (loevgaard) [reply src/Resources/views/admin/order/update/_order_items.html.twig]: Yes, I think I also might have said that :D That's also why I am asking a bit defensively :D 

> Anyway, we should probably have a disabled field with variant here, as in other case the form nulls the value on the order item.

Are you sure they are nulled when they are not posted? I don't think so, and it worked before where this widget wasn't there

File: src/Exception/NewOrderWrongTotalException.php (hunk 12 above)
Comment (loevgaard) [reply src/Exception/NewOrderWrongTotalException.php]: Yes, totally agree. I just never saw this before

File: src/Processor/UpdatedOrderProcessor.php (hunk 13 above)
Comment (loevgaard) [reply src/Processor/UpdatedOrderProcessor.php]: Haha, not a reason that wins an argument 😂 I guess Matthias has a few here: https://matthiasnoback.nl/2014/05/inject-the-manager-registry-instead-of-the-entity-manager/, but also here are a few of my own
- I just got accustomed doing it
- The entity manager can be closed and you can reset it
- You _know_ that no matter what magic people have done, you will get the right manager
- I get some typing from Psalm that Sylius doesn't provide yet

As I said, nothing that wins arguments :D

File: src/Resources/config/services/order_processing.xml (hunk 20 above)
Comment (loevgaard) [reply src/Resources/config/services/order_processing.xml]: For now it's easier to keep the old way, because all other services use that :D

But this is something I will change in the future I think :D 

👍

File: src/Processor/UpdatedOrderProcessor.php (hunk 25 above)
Comment (Zales0123) [reply src/Processor/UpdatedOrderProcessor.php]: 👍 I can make a validator that would check the `$initalTotal < $newOrderTotal` condition and leave the entityManager in a controller 👌

Comment (loevgaard) [reply src/Processor/UpdatedOrderProcessor.php]: Yes, that check probably belongs in a validator 👍

File: src/Processor/UpdatedOrderProcessor.php (hunk 13 above)
Comment (Zales0123) [reply src/Processor/UpdatedOrderProcessor.php]: Let me then pause it for now to not make the iteration over this PR longer than necessary, but we can definitely come back to this discussion later, ok? :)

File: src/Resources/views/admin/order/update/_order_items.html.twig (hunk 6 above)
Comment (Zales0123) [reply src/Resources/views/admin/order/update/_order_items.html.twig]: Zrzut ekranu 2024-06-14 o 12 01 27

I've checked it one more time, and the variant is nulled :dancer: but we can switch to `PATCH` method and then it should be ok, wdyt?

File: src/Processor/UpdatedOrderProcessor.php (hunk 13 above)
Comment (loevgaard) [reply src/Processor/UpdatedOrderProcessor.php]: Yes, of course. Let's move on

File: src/Resources/views/admin/order/update/_order_items.html.twig [hunk 26]
```diff
@@ -28,7 +28,7 @@
                 {% set orderItem = itemForm.vars.data %}
                 <tr>
                     <td>{{ form_row(itemForm.quantity) }}</td>
-                    <td>{{ orderItem.productName }}</td> {# todo add variant #}
```
Comment (loevgaard) [inline src/Resources/views/admin/order/update/_order_items.html.twig:31]: The todo should not be removed, because we want to output the variant also, so people know exactly what product it is

Comment (Zales0123) [reply src/Resources/views/admin/order/update/_order_items.html.twig:31]: I can just fix it already 😄

Comment (loevgaard) [reply src/Resources/views/admin/order/update/_order_items.html.twig:31]: That's even better 🥇

File: infection.json.dist [hunk 27]
```diff
@@ -2,6 +2,13 @@
     "source": {
         "directories": [
             "src"
+        ],
+        "excludes": [
+            "Controller",
+            "DependencyInjection/SetonoSyliusOrderEditExtension.php",
+            "Exception",
+            "EventSubscriber",
+            "Form"
```
Comment (loevgaard) [inline infection.json.dist]: Instead of this I will just lower the msi's. It's not a problem doing that. This is something we can review by the end of project, e.g. looking over mutants to see if there are some important ones

File: infection.json.dist [hunk 28]
```diff
@@ -12,5 +19,6 @@
         }
     },
     "minMsi": 100.00,
-    "minCoveredMsi": 100.00
+    "minCoveredMsi": 100.00,
+    "testFrameworkOptions": "--testsuite=unit"
```
Comment (loevgaard) [inline infection.json.dist]: Isn't it better to include all phpunit tests in the msi?

File: src/Controller/EditOrderAction.php [hunk 29]
```diff
@@ -0,0 +1,82 @@
...
+            $this->postUpdateChangesChecker->check($oldOrder, $updatedOrder);
+            $this->entityManager->flush();
+        } catch (NewOrderWrongTotalException) {
+            return $this->addFlashAndRedirect(
+                'error',
+                'setono_sylius_order_edit.order_update.total_error',
+                'sylius_admin_order_update',
+                $id,
+            );
+        } catch (\Exception) {
```
Comment (loevgaard) [inline src/Controller/EditOrderAction.php]:
```suggestion
        } catch (\Throwable) {
```

File: src/Controller/EditOrderAction.php [hunk 30]
```diff
@@ -0,0 +1,82 @@
...
+    public function __invoke(Request $request, int $id): Response
+    {
+        $order = $this->oldOrderProvider->prepareToUpdate($id);
+
+        $oldOrder = clone $order;
+        $updatedOrder = $this->updatedOrderProvider->provideFromOldOrderAndRequest($order, $request);
+
+        try {
+            $this->updatedOrderProcessor->process($updatedOrder);
+            $this->postUpdateChangesChecker->check($oldOrder, $updatedOrder);
```
Comment (loevgaard) [inline src/Controller/EditOrderAction.php:42]: What's the reason for not doing this as a form validator?

File: src/Processor/UpdatedOrderProcessor.php [hunk 31]
```diff
@@ -0,0 +1,35 @@
...
+        $newOrder->setState(OrderInterface::STATE_NEW);
+        $this->orderInventoryOperator->hold($newOrder);
+
+        return $newOrder;
+//        if ($initialTotal < $newOrder->getTotal()) {
+//            throw new NewOrderWrongTotalException();
+//        }
+//
+//        $this->entityManager->flush();
+    }
```
Comment (loevgaard) [inline src/Processor/UpdatedOrderProcessor.php]: Remove

File: src/Processor/UpdatedOrderProcessorInterface.php [hunk 32]
```diff
@@ -0,0 +1,12 @@
...
+
+declare(strict_types=1);
+
+namespace Setono\SyliusOrderEditPlugin\Processor;
+
+use Sylius\Component\Core\Model\OrderInterface;
+
+interface UpdatedOrderProcessorInterface
+{
+    public function process(OrderInterface $newOrder): OrderInterface;
```
Comment (loevgaard) [inline src/Processor/UpdatedOrderProcessorInterface.php]:
```suggestion
    public function process(OrderInterface $order): OrderInterface;
```

or

```suggestion
    public function process(OrderInterface $updatedOrder): OrderInterface;
```

to be consistent with the service name.

File: infection.json.dist (hunk 27 above)
Comment (Zales0123) [reply infection.json.dist]: On the other hand, these are directories that would not be directly covered with tests. Unit testing forms, event subscribers, or exceptions is redundant, as they're well covered by functional tests, so from my perspective it's better to explicitly show we don't want them to be mutated. I'm open for the discussion, obviously 🖖

File: infection.json.dist (hunk 28 above)
Comment (Zales0123) [reply infection.json.dist]: My understanding of mutation tests is it's great for checking the edge cases not covered by unit tests - the functional tests would reflect the business value of the application and does not have to cover all the edge cases, as they could be covered on the unit level. And are just much more demanding, pretty heavy to mutate them. Again, as always, I'm open for the discussion 🖖

File: src/Controller/EditOrderAction.php (hunk 30 above)
Comment (Zales0123) [reply src/Controller/EditOrderAction.php:42]: As it's a business logic check, I feel it's just easier from the DX perspective to use a custom service with a custom exception. Also validator enforces on us some contract that seems irrelevant for this vase (validation groups, separate classes for constraints) and could we even easily pass two values to the validator (old order and updated order)? 
Again, as most of the conversations we have here, it's a matter of taste and conventions, not a real problem from the application point of view 😄 If you want me to change it to the validator - let me know 🖖

File: infection.json.dist (hunk 27 above)
Comment (loevgaard) [reply infection.json.dist]: Now that the functional tests are written with PHPUnit and Infection runs PHPUnit by default, my take is that we can just as well include everything in infection and just lower the MSI accordingly. This will provide insights later on as to where our tests are lacking. By excluding, you won't get these insights from Infection.

File: infection.json.dist (hunk 28 above)
Comment (loevgaard) [reply infection.json.dist]: Infection mutates the code, not the tests, and I really believe we shouldn't limit Infection just to increase the MSI. I would much rather have a low MSI that tries to mutate the whole code base, so that we can see where we have the 'worst' edge cases

File: src/Controller/EditOrderAction.php (hunk 30 above)
Comment (loevgaard) [reply src/Controller/EditOrderAction.php:42]: There's a lot of problems with doing it using the validator, I agree. Let's keep it this way

File: infection.json.dist (hunk 27 above)
Comment (Zales0123) [reply infection.json.dist]: Ok, let's do it your way, I will lower the MSI and let's see how it goes 🫡

File: infection.json.dist (hunk 28 above)
Comment (Zales0123) [reply infection.json.dist]: [OK](https://github.com/Setono/sylius-order-edit-plugin/pull/3#discussion_r1639809054) :)

File: .github/workflows/build.yaml [hunk 33]
```diff
@@ -316,7 +339,7 @@ jobs:
                     dependency-versions: "${{ matrix.dependencies }}"
 
             -   name: "Collect code coverage with pcov and phpunit/phpunit"
-                run: "vendor/bin/phpunit --coverage-clover=.build/logs/clover.xml"
+                run: "vendor/bin/phpunit tests/Unit/ --coverage-clover=.build/logs/clover.xml"
```
Comment (loevgaard) [inline .github/workflows/build.yaml]:
```suggestion
                run: "vendor/bin/phpunit --coverage-clover=.build/logs/clover.xml"
```

I think this should also include the functional tests. Don't you?

**Review by loevgaard — APPROVED** [review]
